src/OpticalFlowEstimator.cpp

```diff
@@ -0,0 +1,598 @@
+        cv::gpu::GpuMat scaled_image;
+        cv::gpu::resize(last_scaled_image_,
+                        scaled_image,
+                        target_size_);
+        last_scaled_image_ = scaled_image;
+
+        cv::gpu::GpuMat scaled_grayscale_image;
+        cv::gpu::cvtColor(last_scaled_image_,
+                          scaled_grayscale_image,
+                          CV_RGBA2GRAY);
+        last_scaled_grayscale_image_ = scaled_grayscale_image;
+    }
+
+    return true;
+}
+
+void OpticalFlowEstimator::update(const sensor_msgs::Image::ConstPtr& message)
+{
+    // initial time for measurement
+    const int64 start = cv::getTickCount();
+
+    // make sure our current position is up to date
+    if (!updateFilteredPosition(
+                message->header.stamp,
+                ros::Duration(flow_estimator_settings_.tf_timeout))) {
+        ROS_ERROR("Unable to update position for optical flow");
+        return;
+    }
+
+    if (debug_settings_.debug_times) {
+        ROS_WARN("updateFilteredPosition: %f",
+                 (cv::getTickCount() - start) / cv::getTickFrequency());
+    }
+
+    if (current_altitude_ < flow_estimator_settings_.min_estimation_altitude) {
+        ROS_WARN("Height (%f) is below min processing height (%f)",
+                 current_altitude_,
+                 flow_estimator_settings_.min_estimation_altitude);
+        return;
+    }
+
+    const boost::shared_ptr<const cv_bridge::CvImage> curr_image_msg
+        = cv_bridge::toCvShare(message);
+    const cv::Mat& curr_image = curr_image_msg->image;
+
+    if (have_valid_last_image_) {
+        if (curr_image.size() != expected_input_size_) {
+            ROS_ERROR("Ignoring image of size (%dx%d), expected (%dx%d)",
+                      curr_image.size().width,
+                      curr_image.size().height,
+                      expected_input_size_.width,
+                      expected_input_size_.height);
+            return;
+        }
+
+        try {
+            const int64 start = cv::getTickCount();
+
+            // Scale and convert input image
+            cv::gpu::GpuMat d_frame1_big(curr_image);
+            cv::gpu::GpuMat scaled_image;
+            cv::gpu::GpuMat scaled_gray_image;
+
+            cv::gpu::resize(d_frame1_big,
+                            scaled_image,
+                            target_size_);
+
+            if (debug_settings_.debug_times) {
+                ROS_WARN("post resize: %f",
+                         (cv::getTickCount() - start) / cv::getTickFrequency());
+            }
+
+            cv::gpu::cvtColor(scaled_image,
+                              scaled_gray_image,
+                              CV_RGBA2GRAY);
+
+            if (debug_settings_.debug_times) {
+                ROS_WARN("post cvtColor: %f",
+                         (cv::getTickCount() - start) / cv::getTickFrequency());
+            }
+
+            // Get velocity estimate from average vector
+            processImage(scaled_image,
```

Comment:
I think the image resizing should occure in processImage if possible or separate function.

Reply:
Fixed